### DISCUSSION TITLE
Making a generic router

### DISCUSSION THREAD

**Discussion by willburden** [tokio-rs/axum#2184] [Q&A] [answered]
I'm trying to create a generic function (`resource_router`) that returns a Router, so I can eventually have something like:

```rust
// Where Account and Post are structs that implement Resource:

fn router() -> Router {
    Router::new()
        .nest("/accounts", resource_router::<Account>)
        .nest("/posts", resource_router::<Post>)
}
```

Here is a minimal reproduction of the error I'm encountering:

```rust
use axum::{Router, routing::get, Json};
use serde::{de::DeserializeOwned, Deserialize};

trait Resource: DeserializeOwned {}

#[derive(Deserialize)]
struct Account {
    email: String,
}

impl Resource for Account {}

fn resource_router<R: Resource>() -> Router {
    Router::new().route("/", get(get_resource::<R>))
}

async fn get_resource<R: Resource>(Json(_payload): Json<R>) {
    todo!()
}
```

And the error message:

```
error[E0277]: the trait bound `fn(axum::Json<R>) -> impl std::future::Future<Output = ()> {routes::get_resource::<R>}: axum::handler::Handler<_, _, _>` is not satisfied
   --> src/routes.rs:13:34
    |
13  |     Router::new().route("/", get(get_resource::<R>))
    |                              --- ^^^^^^^^^^^^^^^^^ the trait `axum::handler::Handler<_, _, _>` is not implemented for fn item `fn(axum::Json<R>) -> impl std::future::Future<Output = ()> {routes::get_resource::<R>}`
    |                              |
    |                              required by a bound introduced by this call
    |
    = help: the following other types implement trait `axum::handler::Handler<T, S, B>`:
              <axum::handler::Layered<L, H, T, S, B, B2> as axum::handler::Handler<T, S, B2>>
              <axum::routing::MethodRouter<S, B> as axum::handler::Handler<(), S, B>>
note: required by a bound in `axum::routing::get`
   --> /home/will/.cargo/registry/src/index.crates.io-6f17d22bba15001f/axum-0.6.20/src/routing/method_routing.rs:403:1
    |
403 | top_level_handler_fn!(get, GET);
    | ^^^^^^^^^^^^^^^^^^^^^^---^^^^^^
    | |                     |
    | |                     required by a bound in this function
    | required by this bound in `get`
    = note: this error originates in the macro `top_level_handler_fn` (in Nightly builds, run with -Z macro-backtrace for more info)
```

I tried using `axum_macros::debug_handler` but discovered that it doesn't support generic functions. I believe that my handler here satisfies all the requirements listed in the documentation:

> - Are async fns.
> - Take no more than 16 arguments that all implement [FromRequest](https://docs.rs/axum/latest/axum/extract/trait.FromRequest.html).
> - Returns something that implements [IntoResponse](https://docs.rs/axum/latest/axum/response/trait.IntoResponse.html).
> - If a closure is used it must implement Clone + Send and be 'static.
> - Returns a future that is Send. The most common way to accidentally make a future !Send is to hold a !Send type across an await.

I've found that the error vanishes if I make any of the following changes:

- Make `resource_router` use a specific `Resource`:
  ```rust
  fn resource_router() -> Router {
      Router::new().route("/", get(get_resource::<Account>))
  }
  ```
- Remove the `Json` extractor from the handler:
  ```rust
  async fn get_resource<R: Resource>() {
      todo!()
  }
  ```

So my guess is that the Json extractor isn't meeting the `FromRequest` trait bound; it's as though `Resource` doesn't have a `Deserialize` trait bound, even though it does.

Any help would be greatly appreciated.

**Comment by davidpdrsn** [comment, marked as answer]
You need to also make sure that `R` is `Send + 'static`:

```rust
trait Resource: DeserializeOwned + Send + 'static {}

fn resource_router<R: Resource>() -> Router {
    Router::new().route("/", get(get_resource::<R>))
}

async fn get_resource<R: Resource>(_: Json<R>) {}
```

The `Send` bound comes from the impl for `Handler` for functions that take one argument:

![one](https://github.com/tokio-rs/axum/assets/718941/58bbf02c-098c-4996-9f95-8453a198c536)

and the `'static` bound comes from `get`:
![two](https://github.com/tokio-rs/axum/assets/718941/e84d3acc-5a8c-43ac-a79a-39548732729d)
